opm/grid/cpgrid/CpGrid.cpp

```diff
@@ -1914,12 +1932,37 @@ bool CpGrid::adapt()
     const std::vector<std::array<int,3>>& cells_per_dim_vec = {{2,2,2}}; // Arbitrary chosen values.
     std::vector<int> assignRefinedLevel(current_view_data_-> size(0));
     const auto& preAdaptMaxLevel = this ->maxLevel();
+    int local_marked_elem_count = 0;
     for (int elemIdx = 0; elemIdx < current_view_data_->size(0); ++elemIdx) {
         const auto& element = cpgrid::Entity<0>(*current_view_data_, elemIdx, true);
         assignRefinedLevel[elemIdx] = (this->getMark(element) == 1) ? (preAdaptMaxLevel +1) : 0;
+        if (this->getMark(element) == 1) {
+            ++local_marked_elem_count;
+        }
     }
-    const std::vector<std::string>& lgr_name_vec = { "LGR" + std::to_string(preAdaptMaxLevel +1) };
-    return this-> adapt(cells_per_dim_vec, assignRefinedLevel,lgr_name_vec);
+    // Check if its a global refinement
+    bool is_global_refine = false;
+    std::vector<std::string> lgr_name_vec = { "LGR" + std::to_string(preAdaptMaxLevel +1) };
+    // Rewrite if global refinement
+#if HAVE_MPI
+    auto global_marked_elem_count = comm().sum(local_marked_elem_count);
+    auto global_cell_count_before_adapt = comm().sum(current_view_data_-> size(0)); // Recall overlap cells are also marked
+    if (global_marked_elem_count == global_cell_count_before_adapt) {
+        // GR stands for GLOBAL REFINEMET
+        lgr_name_vec = { "GR" + std::to_string(preAdaptMaxLevel +1) };
+        is_global_refine = true; // parallel
+    }
+#endif
+    if (local_marked_elem_count == current_view_data_-> size(0)) {
+        // GR stands for GLOBAL REFINEMET
+        lgr_name_vec = { "GR" + std::to_string(preAdaptMaxLevel +1) };
+        is_global_refine = true; // sequential
+    }
```

Comment:
The old version had a previous check for MPI above these lines. I see that it's not enough only to check this after that. In any case, the first suggestion was to remove this check from here and guarantee that the "GR" is set when someone calls global refinement.

In my humble opinion, I would prefer to fix the check here. When the check is correct, then calling adapt having marked all the elements will also end up being called "GR" instead of "LGR". 

It might be nice that the label of the refinement/adaptation reflects if it's a local or a global one, independently of having called globalRefine(). What do you think?

Reply:
Thanks.